src/OpenFeature/Api.cs

```diff
@@ -36,31 +35,26 @@ public sealed class Api
         private Api() { }
 
         /// <summary>
-        /// Sets the feature provider
+        /// Sets the feature provider. In order to wait for the provider to be set, and initialization to complete,
+        /// await the returned task.
         /// </summary>
+        /// <remarks>The provider cannot be set to null. Attempting to set the provider to null has no effect.</remarks>
         /// <param name="featureProvider">Implementation of <see cref="FeatureProvider"/></param>
-        public void SetProvider(FeatureProvider featureProvider)
+        public async Task SetProvider(FeatureProvider featureProvider)

```

Comment:
So, returning a task here is a breaking API change. This seems like the most appropriate action. Await the method if you care that it finishes initialization, don't await if you do not.

Other options would be adding another method, but this method would just be async without a way to know it finished, versus being sync.

This all applies to named providers as well.

Reply:
Yea I've been thinking about this myself its quite common to offer bother sync and async libraries especially when working on a old legacy project in full framework. Lets just leave it for now and see if anyone raise concern before going ahead an introduce a sync based methods.

Reply:
> So, returning a task here is a breaking API change. This seems like the most appropriate action. Await the method if you care that it finishes initialization, don't await if you do not.
>
> Other options would be adding another method, but this method would just be async without a way to know it finished, versus being sync.

"events" are the mechanism by which you can call this synchronously and "know" the outcome (your READY/ERROR event handlers will run). I know they don't exist on this SDK yet, but they will eventually.

> Other options would be adding another method

This is what's been done in Java and JS, and what the spec recommends: https://openfeature.dev/specification/sections/flag-evaluation/#requirement-118

Reply:
(To elaborate `AndWait` or `Wait` in general have been used to provide sync APIs where the base API is async, so it would be a confusing name)

Reply:
> The spec recommends setProviderAndWait

The spec intentionally doesn't prescribe a name for the function - everything outside of the block quotes is non-normative, and mostly there for example/clarity - we could certainly call them `setProvider` and `setProviderAsync`, which I agree would be optimal.

> (To elaborate AndWait or Wait in general have been used to provide sync APIs where the base API is async, so it would be a confusing name)

I agree, "wait" this implies the method will _actually_ block, in the Thread.Sleep sense, not await some callback/event.

Do you have objections to adding a `setProviderAsync` method instead? I agree that there's really no need for a synchronous `setProvider`, but given that there's equivalents in other SDKs and it would be a major version bump, I'm hesitant to changing it to async.

cc @benjiro

Reply:
Adding `setProviderAsync` is the lesser of two evils. The biggest problem with it is that it creates the implied meaning that `setProvider` is synchronous. We will document it, but unless people are thorough that will be the expectation.

To counter this we maybe should mark `setProvider` as obsolete, and then when there is a major version remove it. 

`[Obsolete("setProvider is deprecated, please use setProviderAsync instead. Note that setProvider executes asynchronously. To achieve the same functionality use setProviderAsync and do not wait for the result.")]`

Reply:
I'd OK with either path, though my preference would be marking `setProvider` obsolete to avoid breaking changes.

Let's see what others think.

Reply:
Most dotnet devs are use to the Async suffix convention, so they will expect that the rest of the API follows which could lead to confusion. Usually when this approach is taken a library would provide both async, and sync methods. I think it would be unintuitive to mix the two (non async suffix/async suffix methods returning awaitable tasks). If we are going to go down the path of providing both sync/async methods i don't think there is any clean way without making a breaking change.

Reply:
> Most dotnet devs are use to the Async suffix convention, so they will expect that the rest of the API follows which could lead to confusion. Usually when this approach is taken a library would provide both async, and sync methods. I think it would be unintuitive to mix the two (non async suffix/async suffix methods returning awaitable tasks). If we are going to go down the path of providing both sync/async methods i don't think there is any clean way without making a breaking change.

Well if that's the case, it's breaking either way.

So I guess the question is, is it better to release a 2.0 that supports both (with -Async suffixes) or should we just break this one method and make it async? Or is some 3rd way better?

Reply:
I think, if the API is fully async, then we don't need to worry about the Async suffix. It really is a case of libraries that implemented things synchronously, and then needed async methods as well. 

The interface for this API is almost all Async, and it happened that there was this sync method, so making it async would provide a consistent async interface. 

It may be worth checking the rest of the API to see if other items should move to async with this.

Reply:
(I guess the client is really where things are async, and the "Api" doesn't have that many methods that are async, or would conceivably be async.)

Reply:
I'm fine with keeping it all async and not providing a sync api. I think since we aren't 1.x.x yet its the best time to make a breaking change to SetProvider, i think we just need to call it out in the release notes.

Reply:
This SDK is already version `1.3.1` but we can release a version `2.0` if it makes sense. I think it does make sense in this case since it simplifies waiting for a provider to initialize and the migration process should be straightforward.

Reply:
I'm fine with the breaking change, especially if we agree we'll do the same breaking change in JS, which has the same async/await semantics.